## Train a model and predict activities

The file works you through training a model to detect activities of a given device. The activities can be defined any action a device may allow users to do, and each activity should contain at least 3 repeated experiments to make representative learnings. 

**Before you go ahead, download the required pcap files** from [Google Drive > iot-model.tgz](https://drive.google.com/open?id=1lMqZ5qx6ATqIIiLOdTYcSm6RliK1F7vA) (size=~127MB), and decompress it to the current folder. You should expect the file structure to be `traffic/us/yi-camera/{activity_name}/{datetime}.{length}.pcap`.

**IMPORTANT** Make sure to use `python3` and install all the dependencies. 
- `pip install -r requirements.txt`


#### Extract pcap files to per-flow level info 
(output too long, hidden from this file)

In [1]:
!./raw2intermediate.sh list_exp.txt tagged-intermediate/us

traffic/us//yi-camera/power/2019-04-25_19:28:58.154s.pcap
tagged-intermediate/us/yi-camera/power/2019-04-25_19:28:58.154s.txt exists.
traffic/us//yi-camera/power/2019-04-25_19:25:30.155s.pcap
tagged-intermediate/us/yi-camera/power/2019-04-25_19:25:30.155s.txt exists.
traffic/us//yi-camera/power/2019-04-25_19:21:40.166s.pcap
tagged-intermediate/us/yi-camera/power/2019-04-25_19:21:40.166s.txt exists.
traffic/us//yi-camera/local_move/2019-04-25_19:47:16.40s.pcap
tagged-intermediate/us/yi-camera/local_move/2019-04-25_19:47:16.40s.txt exists.
traffic/us//yi-camera/local_move/2019-04-25_19:48:09.53s.pcap
tagged-intermediate/us/yi-camera/local_move/2019-04-25_19:48:09.53s.txt exists.
traffic/us//yi-camera/local_move/2019-04-25_19:46:16.41s.pcap
tagged-intermediate/us/yi-camera/local_move/2019-04-25_19:46:16.41s.txt exists.
traffic/us//yi-camera/android_lan_photo/2019-04-26_22:48:43.35s.pcap
tagged-intermediate/us/yi-camera/android_lan_photo/2019-04-26_22:48:43.35s.txt exists.
traffic/us//yi-c

tagged-intermediate/us/yi-camera/android_lan_recording/2019-04-26_21:39:55.40s.txt exists.
traffic/us//yi-camera/android_lan_recording/2019-04-26_22:46:16.40s.pcap
tagged-intermediate/us/yi-camera/android_lan_recording/2019-04-26_22:46:16.40s.txt exists.
traffic/us//yi-camera/android_lan_recording/2019-04-26_22:01:10.39s.pcap
tagged-intermediate/us/yi-camera/android_lan_recording/2019-04-26_22:01:10.39s.txt exists.
traffic/us//yi-camera/android_lan_recording/2019-04-26_22:10:05.41s.pcap
tagged-intermediate/us/yi-camera/android_lan_recording/2019-04-26_22:10:05.41s.txt exists.
traffic/us//yi-camera/android_lan_recording/2019-04-26_22:17:20.40s.pcap
tagged-intermediate/us/yi-camera/android_lan_recording/2019-04-26_22:17:20.40s.txt exists.
traffic/us//yi-camera/android_lan_recording/2019-04-26_21:29:34.34s.pcap
tagged-intermediate/us/yi-camera/android_lan_recording/2019-04-26_21:29:34.34s.txt exists.
traffic/us//yi-camera/android_lan_recording/2019-04-26_22:26:26.40s.pcap
tagged-intermedi

tagged-intermediate/us/yi-camera/android_lan_watch/2019-04-26_22:11:23.32s.txt exists.
traffic/us//yi-camera/android_lan_watch/2019-04-26_21:53:33.32s.pcap
tagged-intermediate/us/yi-camera/android_lan_watch/2019-04-26_21:53:33.32s.txt exists.
traffic/us//yi-camera/android_lan_watch/2019-04-26_22:24:04.32s.pcap
tagged-intermediate/us/yi-camera/android_lan_watch/2019-04-26_22:24:04.32s.txt exists.
traffic/us//yi-camera/android_lan_watch/2019-04-26_22:43:58.31s.pcap
tagged-intermediate/us/yi-camera/android_lan_watch/2019-04-26_22:43:58.31s.txt exists.
traffic/us//yi-camera/android_lan_watch/2019-04-26_22:45:45.31s.pcap
tagged-intermediate/us/yi-camera/android_lan_watch/2019-04-26_22:45:45.31s.txt exists.
traffic/us//yi-camera/android_lan_watch/2019-04-26_22:36:53.32s.pcap
tagged-intermediate/us/yi-camera/android_lan_watch/2019-04-26_22:36:53.32s.txt exists.
traffic/us//yi-camera/android_lan_watch/2019-04-26_22:04:11.30s.pcap
tagged-intermediate/us/yi-camera/android_lan_watch/2019-04-26_22

tagged-intermediate/us/yi-camera/android_wan_watch/2019-04-27_22:24:58.33s.txt exists.
traffic/us//yi-camera/android_wan_watch/2019-04-27_22:15:53.32s.pcap
tagged-intermediate/us/yi-camera/android_wan_watch/2019-04-27_22:15:53.32s.txt exists.
traffic/us//yi-camera/android_wan_watch/2019-04-27_21:57:42.32s.pcap
tagged-intermediate/us/yi-camera/android_wan_watch/2019-04-27_21:57:42.32s.txt exists.
traffic/us//yi-camera/android_wan_watch/2019-04-27_21:52:16.32s.pcap
tagged-intermediate/us/yi-camera/android_wan_watch/2019-04-27_21:52:16.32s.txt exists.
traffic/us//yi-camera/android_wan_watch/2019-04-27_22:06:48.32s.pcap
tagged-intermediate/us/yi-camera/android_wan_watch/2019-04-27_22:06:48.32s.txt exists.
traffic/us//yi-camera/android_wan_watch/2019-04-27_21:48:39.32s.pcap
tagged-intermediate/us/yi-camera/android_wan_watch/2019-04-27_21:48:39.32s.txt exists.
traffic/us//yi-camera/android_wan_watch/2019-04-27_21:55:54.32s.pcap
tagged-intermediate/us/yi-camera/android_wan_watch/2019-04-27_21

tagged-intermediate/us/yi-camera/android_wan_recording/2019-04-27_21:29:09.39s.txt exists.
traffic/us//yi-camera/android_wan_recording/2019-04-27_22:25:31.41s.pcap
tagged-intermediate/us/yi-camera/android_wan_recording/2019-04-27_22:25:31.41s.txt exists.
traffic/us//yi-camera/android_wan_recording/2019-04-27_22:40:05.41s.pcap
tagged-intermediate/us/yi-camera/android_wan_recording/2019-04-27_22:40:05.41s.txt exists.
traffic/us//yi-camera/android_wan_recording/2019-04-27_22:36:30.40s.pcap
tagged-intermediate/us/yi-camera/android_wan_recording/2019-04-27_22:36:30.40s.txt exists.
traffic/us//yi-camera/android_wan_recording/2019-04-27_22:03:43.39s.pcap
tagged-intermediate/us/yi-camera/android_wan_recording/2019-04-27_22:03:43.39s.txt exists.
traffic/us//yi-camera/android_wan_recording/2019-04-27_22:07:20.41s.pcap
tagged-intermediate/us/yi-camera/android_wan_recording/2019-04-27_22:07:20.41s.txt exists.
traffic/us//yi-camera/android_wan_recording/2019-04-27_22:12:46.40s.pcap
tagged-intermedi

#### Parse per-flow info to features per-activity
(output too long, hidden from this file)

In [2]:
!python extract_tbp_features.py

/bin/sh: 1: python: not found


#### Train the model using the features
(re-reun the command below will skip the trained model, delete .model and .label.txt file to re-train) 

In [4]:
!python3 train_rf_models.py

/home/amenon/.local/lib/python3.6/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
mkdir: created directory 'tagged-models'
mkdir: created directory 'tagged-models/us'
mkdir: created directory 'tagged-models/us/output'
Scanning features/us/yi-camera.csv
  #Data points: 2490 
	Variable: spanOfGroup          Importance: 0.409
	Variable: q90                  Importance: 0.067
	Variable: meanTBP              Importance: 0.065
	Variable: kurtosisLength       Importance: 0.048
	Variable: q60                  Importance: 0.048
	Variable: q80                  Importance: 0.048
	Variable: meanBytes            Importance: 0.042
	Variable: q70                  Importance: 0.039
	Variable: medAbsDev            Importance: 0.036
	Variable: q40                  Importance: 0.035
	Variable: skewLength           Importance: 0.03
	Varia

#### Predict activities given a pcap file

`Usage: ./predict.sh device_name path-to-pcap result-file modeldir
    Note that a temprary file  /tmp/{md5}.txt will be created during the process
    Requires python3`

In [4]:
!./predict.sh yi-camera sample-yi-camera-recording.pcap sample-result.csv tagged-models/us/

/tmp/3b986fce33742216b5f4d8d2e427766d.txt
/tmp/3b986fce33742216b5f4d8d2e427766d.txt exists, delete it to re-parse!
/Users/jingjing/.pyenv/versions/3.6.0/lib/python3.6/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py:47: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
tagged-models/us//yi-camera.model
Total packets: 1621
Number of slices: 2
Result:
             ts        ts_end  ts_delta  num_pkt              state
0  1.556329e+09  1.556329e+09  0.000019     1620  android_lan_watch
Results saved to sample-result.csv


In [5]:
!cat sample-result.csv

ts,ts_end,ts_delta,num_pkt,state,device
1556329377.198794,1556329407.828307,1.9e-05,1620,android_lan_watch,yi-camera


Explanation - between epoch time 1556329377.198794 and 1556329407.828307, the network traffic from yi-camera was predicted as the same activity as android_lan_watch, which is using android companion app to watch the video from the camera when both devices are connected to the same WI-FI network.